src/transport/SecureSession.h

```diff
@@ -249,6 +252,11 @@ class SecureSession : public Session, public ReferenceCounted<SecureSession, Sec
         // grab this session, when all SessionHandles go out of scope, the
         // session object will be released automatically.
         kPendingRemoval = 3,
+
+        // For UpdateNOC command, the session still functional, but it can't
+        // yield any new exchanges, the last exchange running UpdateNOC command
+        // will close the session soon.
+        kInactive = 4,
```

Comment:
IMHO the solution here feels more involved to me compared to the exchange-pinning approach. In this solution, both the session and exchange now have to be aware of this special behavior around UpdateNOC, while the former just affects the ExchangeContext.

Importantly, this also starts to really make `SessionHolder` behave both as a strong and a weak-ref, which I think, makes its role confusing. Having `SessionHolder` just be a weak-ref style entity when holding onto a SecureSession was the easy to intuit idea. With this, it now clearly functions as a strong reference too to hold on to the session, which I think, only adds to the mental complexity of this already complicated space.

Reply:
> this also starts to really make SessionHolder behave both as a strong and a weak-ref

This is not true, the SessionHolder behaves exactly as same as which it is in Active state. It is not changed that which strong-ref is holding the session.

Reply:
> What happens if our special session in the Inactive state gets evicted?

It just get evicted, along with the special exchange, nothing special here. If we choose to evict it, we did our best effort to replay `UpdateNOC` command. This aligns with the spec.

> Can it be evicted?

Yes, it can, there is nothing special with the session.

Reply:
If that's the setup, then I have to admit I like not having to have non-evictable sessions...  I'm still not sure about having special eviction code.